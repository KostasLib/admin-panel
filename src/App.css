.panel {
  width: 100vw;
  height: 100vh;
  background-color: white;
  margin: auto;

  /* Grid */
  display: grid;
  grid-template-columns: 7rem auto;
}
@media screen and (min-width: 768px) {
  .panel {
    width: 80vw;
    height: 60vh;
    box-shadow: 0 0 5px 0 lightgray;

    /* Grid */
    grid-template-columns: 1fr 1fr;
  }

  .list-item {
    grid-template-columns: 7rem 0;
  }
}

.panel__left {
  overflow: auto;
}

.panel__right {
  background-color: lightcoral;
  overflow: auto;
}

.list-item {
  height: 6rem;
  display: grid;
  grid-template-columns: 7rem auto;
}

.list-item:hover {
  background-color: #e8e8e8;
}
