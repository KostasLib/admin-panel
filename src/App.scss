@import 'variables';

.panel {
  // Styles
  width: 100vw;
  height: 100vh;
  background-color: white;
  margin: auto;

  // Flex
  display: flex;

  // Fullscreen on mobile, smaller on desktop. Width can be reduced up to the breakpoint's size.
  @media (min-width: $bp-desktop) {
    min-width: $bp-desktop;
    width: 60vw;
    height: 60vh;
    box-shadow: 0 0 5px 1px lightgray;
  }

  &__section {
    flex: 1;
    overflow: auto;
  }

  &__squeezeable-section {
    overflow: auto;
    // Take up min space on mobile and grow on desktop
    @media (min-width: $bp-desktop) {
      flex: 1;
    }
  }
}
