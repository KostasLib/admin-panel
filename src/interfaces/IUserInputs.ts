export interface IUSerInputs {
  name: string;
  email: string;
  phone: string;
  address: string;
  company: string;
}
