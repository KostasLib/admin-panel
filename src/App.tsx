import React from 'react';
import './App.css';
import './normalize.css';

const App: React.FC = () => {
  return <View />;
};

export default App;

const View: React.FC = () => {
  return (
    <article className="panel">
      <section className="panel__left">
        <ul>
          {Array(20)
            .fill(0)
            .map((el, i) => (
              <li key={i} className="list-item">
                hello {i}
              </li>
            ))}
        </ul>
      </section>
      <section className="panel__right">hello from right</section>
    </article>
  );
};
