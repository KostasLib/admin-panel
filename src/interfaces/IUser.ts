import { IUSerInputs } from './IUserInputs';

export interface IUser extends IUSerInputs {
  id: string;
  photo?: string;
}
