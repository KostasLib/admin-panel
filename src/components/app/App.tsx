import React from 'react';
import { USERS } from '../../user_data';
import { IUSerInputs } from '../../interfaces/IUserInputs';
import { IUser } from '../../interfaces/IUser';
import { View } from './View';

interface IState {
  selected?: string;
  users: IUser[];
}
interface IProps {
  users?: IUser[];
}
/**
 * A container component holidng the App's state.
 * Handles selecting a user and provides a save user service.
 */
class App extends React.Component<IProps, IState> {
  state: IState = {
    selected: undefined,
    users: this.props.users || USERS,
  };
  /**
   * Changes the selected user id.
   * Provides a way to select a user and render based on the selected user.
   */
  handleSelectUser = (id: string) => () => this.setState({ selected: id });

  /** Saves changes made to a user to the App's state. */
  handleSave = (newData: IUSerInputs) => {
    this.setState(prev => ({
      users: prev.users.map(user =>
        user.id === prev.selected ? { ...user, ...newData } : user,
      ),
    }));
  };

  render() {
    return (
      <View
        users={this.state.users}
        onSelectUser={this.handleSelectUser}
        selected={this.state.selected}
        onSave={this.handleSave}
      />
    );
  }
}

export default App;
