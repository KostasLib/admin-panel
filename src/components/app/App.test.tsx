import React from 'react';
import App from './App';
import { shallow } from 'enzyme';
import { IUser } from '../../interfaces/IUser';
import { IUSerInputs } from '../../interfaces/IUserInputs';

describe('The <App/> component', () => {
  const users: IUser[] = [
    {
      id: '5c093af1c6ee9117a581c7d6',
      photo: 'https://randomuser.me/api/portraits/men/40.jpg',
      name: 'Bates Washington',
      company: 'ZOLAREX',
      email: 'bates.washington@zolarex.io',
      phone: '+1 (915) 447-2207',
      address: '958 Brevoort Place, Ona, Maine, 2433',
    },
    {
      id: '5c093af1aeca1bb00607fb2a',
      photo: 'https://randomuser.me/api/portraits/men/1.jpg',
      name: 'Mollie Oneill',
      company: 'VIAGRAND',
      email: 'mollie.oneill@viagrand.biz',
      phone: '+1 (852) 535-3880',
      address: '120 Cedar Street, Mansfield, Kentucky, 8890',
    },
  ];

  const inputs: IUSerInputs = {
    name: '',
    email: '',
    address: '',
    company: '',
    phone: '',
  };

  it('should render only the View component', () => {
    const component = shallow(<App users={users} />);
    expect(component).toMatchSnapshot();
    expect(component.length).toBe(1);
  });

  it('initially has no user selected', () => {
    const component = shallow(<App users={users} />);
    expect(component.state('selected')).toBeUndefined();
  });

  it('should provide a method to select a user and re-render the View with a selected user', () => {
    const { id } = users[0];
    const component = shallow(<App users={users} />);
    const instance = component.instance() as App;

    // Assert there is no initial selected user.
    expect(component).toMatchSnapshot();
    expect(component.state('selected')).toBeUndefined();

    // Call the handleSelectUSer method with an id.
    instance.handleSelectUser(id)();

    // Asssert that a user is now selected.
    expect(component.state('selected')).toBe(id);
    expect(component).toMatchSnapshot();
  });

  it("has a handleSave method that updates a user's data", () => {
    const { id } = users[0];
    const component = shallow(<App users={users} />);
    const instance = component.instance() as App;
    const updated = users.map(u => (u.id === id ? { ...u, ...inputs } : u));

    // Select a user.
    instance.setState({ selected: id });
    expect(component.state('selected')).toBe(id);

    // Call the handleSave method and update the user's properties.
    // Assert the users state is updated.
    instance.handleSave(inputs);
    expect(component.state('users')).toEqual(updated);
  });

  it("the handleSave method should do nothing if there is no selected user or it can't find a user", () => {
    const id = 'user-id';
    const component = shallow(<App users={users} />);
    const instance = component.instance() as App;

    // Test initial state.
    expect(component.state('selected')).toBeUndefined();
    expect(component.state('users')).toEqual(users);

    // Call the handleSave method.
    // Assert the initial state hasn't changed as long as the selected id is undefined.
    instance.handleSave(inputs);
    expect(component.state('users')).toEqual(users);

    // Set selected id.
    instance.setState({ selected: id });
    expect(component.state('selected')).toBe(id);

    // Call the handleSave method.
    // Assert the initial users state hasn't change since the user isn't be found.
    instance.handleSave(inputs);
    expect(component.state('users')).toEqual(users);
  });
});
