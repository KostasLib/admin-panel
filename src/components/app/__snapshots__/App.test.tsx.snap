// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <App/> component should provide a method to select a user and re-render the View with a selected user 1`] = `
<View
  onSave={[Function]}
  onSelectUser={[Function]}
  users={
    Array [
      Object {
        "address": "958 Brevoort Place, Ona, Maine, 2433",
        "company": "ZOLAREX",
        "email": "bates.washington@zolarex.io",
        "id": "5c093af1c6ee9117a581c7d6",
        "name": "Bates Washington",
        "phone": "+1 (915) 447-2207",
        "photo": "https://randomuser.me/api/portraits/men/40.jpg",
      },
      Object {
        "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
        "company": "VIAGRAND",
        "email": "mollie.oneill@viagrand.biz",
        "id": "5c093af1aeca1bb00607fb2a",
        "name": "Mollie Oneill",
        "phone": "+1 (852) 535-3880",
        "photo": "https://randomuser.me/api/portraits/men/1.jpg",
      },
    ]
  }
/>
`;

exports[`The <App/> component should provide a method to select a user and re-render the View with a selected user 2`] = `
<View
  onSave={[Function]}
  onSelectUser={[Function]}
  selectedID="5c093af1c6ee9117a581c7d6"
  users={
    Array [
      Object {
        "address": "958 Brevoort Place, Ona, Maine, 2433",
        "company": "ZOLAREX",
        "email": "bates.washington@zolarex.io",
        "id": "5c093af1c6ee9117a581c7d6",
        "name": "Bates Washington",
        "phone": "+1 (915) 447-2207",
        "photo": "https://randomuser.me/api/portraits/men/40.jpg",
      },
      Object {
        "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
        "company": "VIAGRAND",
        "email": "mollie.oneill@viagrand.biz",
        "id": "5c093af1aeca1bb00607fb2a",
        "name": "Mollie Oneill",
        "phone": "+1 (852) 535-3880",
        "photo": "https://randomuser.me/api/portraits/men/1.jpg",
      },
    ]
  }
/>
`;

exports[`The <App/> component should render only the View component 1`] = `
<View
  onSave={[Function]}
  onSelectUser={[Function]}
  users={
    Array [
      Object {
        "address": "958 Brevoort Place, Ona, Maine, 2433",
        "company": "ZOLAREX",
        "email": "bates.washington@zolarex.io",
        "id": "5c093af1c6ee9117a581c7d6",
        "name": "Bates Washington",
        "phone": "+1 (915) 447-2207",
        "photo": "https://randomuser.me/api/portraits/men/40.jpg",
      },
      Object {
        "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
        "company": "VIAGRAND",
        "email": "mollie.oneill@viagrand.biz",
        "id": "5c093af1aeca1bb00607fb2a",
        "name": "Mollie Oneill",
        "phone": "+1 (852) 535-3880",
        "photo": "https://randomuser.me/api/portraits/men/1.jpg",
      },
    ]
  }
/>
`;
